grafast/website/grafserv/servers/nuxt.md

```diff
@@ -40,101 +40,25 @@ import { serv } from "@/server/grafserv/serv";
-
-import httpProxy from "http-proxy";
-
-export default defineNuxtModule({
-  async setup(options, nuxt) {
-    /**
-     * Register websockets in DEVELOPMENT mode.
-     */
-    if (nuxt.options.dev) {
-      // hook called in development only
-      nuxt.hook("listen", (devServer) => {
-        // create a proxy for routing ws to runtime server created in dev plugin
-        const proxy = httpProxy.createProxy({
-          target: {
-            host: "localhost",
-            port: 3100,
-          },
-        });
-        // registering ws on devServer
-        devServer.on("upgrade", (req, socket, head) => {
-          // routing ws by path
-          switch (req.url) {
-            case "/api/graphql":
-              // proxy websocket to runtime server created in dev plugin
-              proxy.ws(req, socket, head);
-              break;
-            default:
-              socket.destroy();
-          }
-        });
-      });
-      // Registering runtime plugin for dev
-      addServerPlugin("@/modules/grafserv/ws-dev");
-    }
-
-    /**
-     * Register websockets in PRODUCTION mode.
-     */
-    if (!nuxt.options.dev)
-      // Registering runtime plugin for production
-      addServerPlugin("@/modules/grafserv/ws");
-  },
-});
-```
-
-and two Nitro plugins (one for dev, and one for prod)
+## Websocket support (need h3@^1.13.0)
 
-```ts title="modules/grafserv/ws-dev.ts"
-import { Server } from "http";
+```ts title="server/api/graphql-ws.ts"
+import { defineWebSocketHandler } from "h3";
 import { serv } from "@/server/grafserv/serv";
 
-// plugin running in DEVELOPMENT (runtime)
-export default defineNitroPlugin(async (nitroApp) => {
-  // Create a server for handling websockets
-  const server = new Server().listen({ port: 3100 }, () =>
-    console.log("Runtime server listening on port 3100"),
-  );
-  // Cleanly close server (on leave or HMR before plugin reload)
-  nitroApp.hooks.hookOnce("close", () => {
-    server.closeAllConnections();
-    server.close((err) =>
-      err
-        ? console.warn("Runtime server wrongly closed", err)
-        : console.log("Runtime server stopped"),
-    );
-  });
-  // Attaching ws to server
-  serv.attachWebsocketsToServer_experimental(server);
-});
+// Create and export the `/api/graphql` websocket handler
+export default defineWebSocketHandler(serv.makeWsHandler());
 ```
 
-```ts title="modules/grafserv/ws.ts"
+If the ws endpoint and the graphql endpoint share the same path:
+
+```ts title="server/api/graphql.ts"
 import { serv } from "@/server/grafserv/serv";
 
-// plugin running in PRODUCTION (runtime)
-export default defineNitroPlugin(async (nitroApp) => {
-  // this hook will be called only once at first http request
-  nitroApp.hooks.hookOnce("request", (event: H3Event) => {
-    const server = event.node.req.socket.server;
-    if (server) {
-      // Attaching ws to server
-      serv.attachWebsocketsToServer_experimental(server);
-    }
-  });
+export default eventHandler({
```

Comment:
Please ensure we have the relevant imports.

Reply:
Also this is by far the more common case: that the URL is the same for websockets and not, so I would move it above the `graphql-ws` one - ideally right next to the `/graphql` endpoint itself.

Reply:
> Please ensure we have the relevant imports.

Nuxt has an automatic import system, no need to import `eventHandler`, `defineEventHandler`, `defineWebSocketHandler` into a nuxt application. But we can import them explicitly for clarity.

> Also this is by far the more common case: that the URL is the same for websockets and not, so I would move it above the `graphql-ws` one - ideally right next to the `/graphql` endpoint itself.

Agree with this 👍

Reply:
If users of nuxt won't be confused by the lack of imports it's fine - again, I'm unfamiliar with that system. I'm personally not a fan of magic/ambient imports - I like to see where the variables/functions I'm referencing come from.

Reply:
I will make these changes too